packages/flutter/lib/src/services/text_editing_delta.dart

```diff
@@ -0,0 +1,370 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/foundation.dart';
+
+import 'text_editing.dart';
+import 'text_input.dart' show TextEditingValue;
+
+TextAffinity? _toTextAffinity(String? affinity) {
+  switch (affinity) {
+    case 'TextAffinity.downstream':
+      return TextAffinity.downstream;
+    case 'TextAffinity.upstream':
+      return TextAffinity.upstream;
+  }
+  return null;
+}
+
+/// Replaces a range of text in the original string with the text given in the
+/// replacement string.
+String _replace(String originalText, String replacementText, int start, int end) {
+  final String textStart = originalText.substring(0, start);
+  final String textEnd = originalText.substring(end, originalText.length);
+  final String newText = textStart + replacementText + textEnd;
+  return newText;
+}
+
+/// A structure representing a granular change that has occurred to the editing
+/// state as a result of text editing.
+///
+/// See also:
+///
+///  * [TextEditingDeltaInsertion], a delta representing an insertion.
+///  * [TextEditingDeltaDeletion], a delta representing a deletion.
+///  * [TextEditingDeltaReplacement], a delta representing a replacement.
+///  * [TextEditingDeltaNonTextUpdate], a delta representing an update to the
+///  selection and/or composing region.
+///  * [TextInputConfiguration], to opt-in your [TextInputClient] to receive
+///  [TextEditingDelta]'s you must set [TextInputConfiguration.enableDeltaModel]
```

Comment:
In general, can we make sure we link to the `[TextInputConfiguration.enableDeltaModel]` more frequently? It may be confusing for developers to find the entry-point into the feature since it is not frequently referred to. We can simply note that the feature only works if this flag is enabled in a few more relevant places.